python/lsst/pipe/tasks/setPrimaryFlags.py

```diff
@@ -20,12 +20,164 @@
 # the GNU General Public License along with this program.  If not,
 # see <http://www.lsstcorp.org/LegalNotices/>.
 #
-import numpy
+import numpy as np
 from lsst.pex.config import Config, Field, ListField
 from lsst.pipe.base import Task
 from lsst.geom import Box2D
 
 
+def getPatchInner(sources, patchInfo):
+    """Set a flag for each source if it is in the innerBBox of a patch
+
+    Parameters
+    ----------
+    sources : `lsst.afw.table.SourceCatalog`
+        A sourceCatalog with pre-calculated centroids.
+        If `key` is not `None` then the results are written into
+        `sources[key]`.
+    patchInfo : `lsst.skymap.PatchInfo`
+        Patch object
+
+    Returns
+    --------
+    isPatchInner: boolean array
+        True for each source that has a centroid
+        in the inner region of a patch.
+    """
+    # Extract the centroid position for all the sources
+    centroidFlag = sources[sources.getCentroidSlot().getFlagKey()]
+    centroidKey = sources.getCentroidSlot().getMeasKey()
+    x = sources[centroidKey.getX()]
+    y = sources[centroidKey.getY()]
+
+    # set inner flags for each source and set primary flags for sources with no children
+    # (or all sources if deblend info not available)
+    innerFloatBBox = Box2D(patchInfo.getInnerBBox())
+    xmin = innerFloatBBox.getMinX()
+    xmax = innerFloatBBox.getMaxX()
+    ymin = innerFloatBBox.getMinY()
+    ymax = innerFloatBBox.getMaxY()
+    inInner = (x >= xmin) & (x <= xmax) & (y >= ymin) & (y <= ymax)
+
+    # When the centroider fails, we can still fall back to the peak, but we don't trust
+    # that quite as much - so we use a slightly smaller box for the patch comparison.
+    shrunkInnerFloatBBox = Box2D(innerFloatBBox)
+    shrunkInnerFloatBBox.grow(-1)
+    xmin = shrunkInnerFloatBBox.getMinX()
+    xmax = shrunkInnerFloatBBox.getMaxX()
+    ymin = shrunkInnerFloatBBox.getMinY()
+    ymax = shrunkInnerFloatBBox.getMaxY()
+    inShrunkInner = (x >= xmin) & (x <= xmax) & (y >= ymin) & (y <= ymax)
+
+    # Flag sources contained in the inner region of a patch
+    isPatchInner = (centroidFlag & inShrunkInner) | (~centroidFlag & inInner)
+    return isPatchInner
+
+
+def getTractInner(sources, tractInfo, skyMap):
+    """Set a flag for each source if it is contained in a given tract
+
+    Parameters
+    ----------
+    sources : `lsst.afw.table.SourceCatalog`
+        A sourceCatalog with pre-calculated centroids.
+        If `key` is not `None` then the results are written into
+        `sources[key]`.
+    tractInfo : `lsst.skymap.TractInfo`
+        Tract object
+    skyMap : `lsst.skymap.BaseSkyMap`
+        Sky tessellation object
+
+    Returns
+    -------
+    isTractInner: `bool` array
+        True for each source that has a centroid in the same
+        tract as `tractInfo`.
```

Comment:
This is hard to define because different skymap types have different definitions of `skyMap.findTract`. So your description is the way that BaseSkyMap defines `findTract`, but most of the other classes that inherit from `BasSkyMap` override that method and use different criteria. Perhaps the most inclusive description would be "True if the `skyMap.findTract` method returns the same tract as `tractInfo`."